### PR TITLE
src: fmtlib 10 fixes

### PR DESCRIPTION
**Pull request by bazaah**
## Summary

So, this is my first attempt at an upstream contribution to Ceph. So If I did something wrong below, please know it was done from ignorance not malice. I have read the SubmittingPatches doc, but be gentle please :).

This change set contains multiple fixes to various fmtlib related issues, found during packaging Ceph v18 for Archlinux (notably: [gcc 13](https://archlinux.org/packages/core/x86_64/gcc/), [fmtlib 10](https://archlinux.org/packages/extra/x86_64/fmt/)). See [this comment](https://github.com/ceph/ceph/pull/53698#issuecomment-1737158401) for details

I _have_ ran `make check` for Archlinux (and built the project successfully), but I'm not sure how well this will translate to the Ceph CI infrastructure, hopefully it's close enough?

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ronen-fr** [comment]
I do appreciate the PR commit message intentions, but if you can:
move the text there into a comment (like this one), and maintain in
the PR message a concise description of what was changed and why. Thanks

File: src/common/LogEntry.h [hunk 1]
```diff
@@ -187,6 +191,8 @@ inline std::ostream& operator<<(std::ostream& out, const clog_type t)
   }
 }
 
+template <> struct fmt::formatter<clog_type> : fmt::ostream_formatter {};
```
Comment (ronen-fr) [inline src/common/LogEntry.h:194]: why not directly where that type was declared, i.e. in ostream_temp.h?

File: src/common/LogEntry.h [hunk 2]
```diff
@@ -196,14 +202,14 @@ inline std::ostream& operator<<(std::ostream& out, const LogEntry& e)
 
 template <> struct fmt::formatter<EntityName> : fmt::formatter<std::string_view> {
   template <typename FormatContext>
-  auto format(const EntityName& e, FormatContext& ctx) {
-    return formatter<std::string_view>::format(e.to_str(), ctx);
+  auto format(const EntityName& e, FormatContext& ctx) const {
+    return fmt::formatter<std::string_view>::format(e.to_str(), ctx);
```
Comment (ronen-fr) [inline src/common/LogEntry.h:206]: I'd suggest moving the two formatters into the fmt namespace (and saving us a problem with
some compiler versions)

File: src/common/LogEntry.h [hunk 3]
```diff
@@ -196,14 +202,14 @@ inline std::ostream& operator<<(std::ostream& out, const LogEntry& e)
...
   }
 };
 
 template <> struct fmt::formatter<LogEntry> : fmt::formatter<std::string_view> {
   template <typename FormatContext>
-  auto format(const LogEntry& e, FormatContext& ctx) {
+  auto format(const LogEntry& e, FormatContext& ctx) const {
```
Comment (ronen-fr) [inline src/common/LogEntry.h:212]: good catch

**Comment by ronen-fr** [comment]
all in all - thanks for the needed fixes. Please see some comments.

**Comment by bazaah** [comment]
> I do appreciate the PR commit message intentions, but if you can: move the text there into a comment (like this one), and maintain in the PR message a concise description of what was changed and why. Thanks

So something like:

- Cut commit message details
- Re-add in a comment on the PR
- Remove _Changelog_ section from PR description?

**Comment by ronen-fr** [comment]
> So something like:
> 
> * Cut commit message details
> * Re-add in a comment on the PR
> * Remove _Changelog_ section from PR description?

Sounds good

**Comment by bazaah** [comment]
Briefly outlined, these are issues fixed:

1. Missing fmt::ostream_formatter definitions for types

	These used to be implicit, but now fmtlib requires that types
	specifically opt into this behavior.
	
	`~` src/common/LogEntry.h: clog_type
	`+` src/include/neorados/RADOS_fmt.hpp: neorados::Object
	`+` src/include/types_fmt.h: shard_id_t

2. Missing format_to() interface for ceph_le<T>
	
	According to the upstream docs, it should be possible to just use the
	explicit cast to the underlying T for formatting, but for whatever
	reason, this doesn't work.
	
	Therefore, define a templated format_to() for ceph_le<T> casting to
	the underlying T, explicitly.
	
	This is a fmtlib 10 feature -- previously format_to() was reserved for
	enums. However, this is the simplest method I can see to avoid needing
	to #include fmtlib headers in this header tree.

	`~` src/include/byteorder.h: ceph_le<T>

3. Make fmt::formatter.format definitions const

	Required since fmtlib 8, but seemingly only became an issue in 10.

	`~` src/common/LogEntry.h: EntityName, LogEntry

4. Explicitly cast from snapid_t to uint64_t for fmt::sprintf calls

	As the fmt::formatter.parse implementation for snapid_t does not
	handle the formatting qualifier used (or any at all).

	I think previously this implicitly was casted to uint64_t anyway, so
	we're just making the behavior explicit.

	`~` src/osd/SnapMapper.cc: snapid_t

File: src/common/LogEntry.h (hunk 1 above)
Comment (bazaah) [reply src/common/LogEntry.h:194]: I did it here for code locality reasons, as the ostream impl is [just above](https://github.com/ceph/ceph/pull/53698/files/53363c65a938b6f0854c4b5f23890f9289a63ec1#diff-ecb790fabfca8053e26d2e6483a62314e6db02c401ff8ab0204a84fc02807db3R176-R192). I can move it, but that will require `#include`ing fmtlib in common/ostream_temp.h and dependents... is that acceptable?

File: src/common/LogEntry.h (hunk 2 above)
Comment (bazaah) [reply src/common/LogEntry.h]: ca16531f92be19075d3ca820adb3a0de95e393b6

**Comment by bazaah** [comment]
jenkins test make check

**Comment by bazaah** [comment]
jenkins test make check arm64

**Comment by bazaah** [comment]
So I think this is ready for a second round.

I'm not sure arm64 refuses to build (something about missing sanitizers?), and I've answered the only outstanding review comment ([link](https://github.com/ceph/ceph/pull/53698#discussion_r1338428479))

**Comment by bazaah** [comment]
Anything more I can do to move this over the finish line?

**Comment by ronen-fr** [comment]
> Anything more I can do to move this over the finish line?

I plan to wait for a couple more days for @adamemerson  or others, then probably send this to Teuthology

**Comment by a16bitsysop** [comment]
i tried these patches on alpine linux and it seems to be missing a formatter needed in fmt 10.1.1

I get this error

```
ninja: job failed: /usr/bin/clang++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/build/src/include -I/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src -isystem /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/build/include -isystem /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/xxHash -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1  -D_LARGEFILE64_SOURCE -O2 -DNDEBUG -DSTATX_INO=0x100U -Wno-deprecated-declarations -Os -DNDEBUG -std=c++20 -fPIC -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -fdiagnostics-color=auto -MD -MT src/mon/CMakeFiles/mon.dir/LogMonitor.cc.o -MF src/mon/CMakeFiles/mon.dir/LogMonitor.cc.o.d -o src/mon/CMakeFiles/mon.dir/LogMonitor.cc.o -c /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.cc
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.cc:49:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.h:26:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/PaxosService.h:20:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/Monitor.h:46:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/common/LogClient.h:19:
/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/common/LogEntry.h:217:12: error: call to 'format_to' is ambiguous
    return format_to(ctx.out(), "{} {} ({}) {} : {} {} {}",
           ^~~~~~~~~
/usr/include/fmt/core.h:1097:64: note: in instantiation of function template specialization 'fmt::formatter<LogEntry>::format<fmt::basic_format_context<fmt::appender, char>>' requested here
    -> decltype(typename Context::template formatter_type<T>().format(
                                                               ^
/usr/include/fmt/core.h:1108:10: note: while substituting deduced template arguments into function template 'has_const_formatter_impl' [with Context = fmt::basic_format_context<fmt::appender, char>, T = LogEntry]
  return has_const_formatter_impl<Context>(static_cast<T*>(nullptr));
         ^
/usr/include/fmt/core.h:1443:38: note: in instantiation of function template specialization 'fmt::detail::has_const_formatter<LogEntry, fmt::basic_format_context<fmt::appender, char>>' requested here
  struct formattable : bool_constant<has_const_formatter<U, Context>() ||
                                     ^
/usr/include/fmt/core.h:1447:39: note: in instantiation of template class 'fmt::detail::arg_mapper<fmt::basic_format_context<fmt::appender, char>>::formattable<const LogEntry, LogEntry>' requested here
  template <typename T, FMT_ENABLE_IF(formattable<T>::value)>
                                      ^
/usr/include/fmt/core.h:1448:33: note: while substituting prior template arguments into non-type template parameter [with T = const LogEntry]
  FMT_CONSTEXPR FMT_INLINE auto do_map(T& val) -> T& {
                                ^~~~~~~~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:1462:63: note: while substituting deduced template arguments into function template 'do_map' [with T = const LogEntry, $1 = (no value)]
  FMT_CONSTEXPR FMT_INLINE auto map(T& val) -> decltype(this->do_map(val)) {
                                                              ^
/usr/include/fmt/core.h:1478:50: note: while substituting deduced template arguments into function template 'map' [with T = const LogEntry, U = (no value), $2 = (no value)]
    type_constant<decltype(arg_mapper<Context>().map(std::declval<const T&>())),
                                                 ^
/usr/include/fmt/core.h:2609:16: note: in instantiation of template type alias 'mapped_type_constant' requested here
      : types_{mapped_type_constant<Args, buffer_context<Char>>::value...},
               ^
/usr/include/fmt/core.h:2740:47: note: in instantiation of member function 'fmt::detail::format_string_checker<char, LogEntry>::format_string_checker' requested here
      detail::parse_format_string<true>(str_, checker(s));
                                              ^
/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.cc:414:52: note: in instantiation of function template specialization 'fmt::basic_format_string<char, const LogEntry &>::basic_format_string<char[4], 0>' requested here
    fmt::format_to(std::back_inserter(log_buffer), "{}\n", le);
                                                   ^
/usr/include/fmt/core.h:2814:17: note: candidate function [with OutputIt = fmt::appender, T = <const utime_t &, const EntityName &, const entity_name_t &, const unsigned long &, const std::basic_string<char> &, const clog_type &, const std::basic_string<char> &>, $2 = 0]
FMT_INLINE auto format_to(OutputIt out, format_string<T...> fmt, T&&... args)
                ^
/usr/bin/../lib/gcc/x86_64-alpine-linux-musl/13.1.1/../../../../include/c++/13.1.1/format:3774:5: note: candidate function [with _Out = fmt::appender, _Args = <const utime_t &, const EntityName &, const entity_name_t &, const unsigned long &, const std::basic_string<char> &, const clog_type &, const std::basic_string<char> &>]
    format_to(_Out __out, format_string<_Args...> __fmt, _Args&&... __args)
    ^
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.cc:49:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.h:22:
In file included from /usr/include/fmt/format.h:49:
/usr/include/fmt/core.h:1576:63: error: implicit instantiation of undefined template 'fmt::detail::type_is_unformattable_for<const LogEntry, char>'
    type_is_unformattable_for<T, typename Context::char_type> _;
                                                              ^
/usr/include/fmt/core.h:1808:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::appender, char>, const LogEntry, 0>' requested here
        data_{detail::make_arg<is_packed, Context>(args)...} {
                      ^
/usr/include/fmt/core.h:1826:10: note: in instantiation of function template specialization 'fmt::format_arg_store<fmt::basic_format_context<fmt::appender, char>, LogEntry>::format_arg_store<const LogEntry>' requested here
  return {args...};
         ^
/usr/include/fmt/core.h:2816:36: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::basic_format_context<fmt::appender, char>, const LogEntry>' requested here
  return vformat_to(out, fmt, fmt::make_format_args(args...));
                                   ^
/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/mon/LogMonitor.cc:414:10: note: in instantiation of function template specialization 'fmt::format_to<std::back_insert_iterator<fmt::basic_memory_buffer<char>>, const LogEntry &, 0>' requested here
    fmt::format_to(std::back_inserter(log_buffer), "{}\n", le);
         ^
/usr/include/fmt/core.h:1554:45: note: template is declared here
template <typename T, typename Char> struct type_is_unformattable_for;
                                            ^
/usr/include/fmt/core.h:1579:3: error: static assertion failed due to requirement 'formattable': Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
  static_assert(
  ^
3 errors generated.
```

**Comment by bazaah** [comment]
I guess `std::format_to` and `fmt::format_to` are ambiguous somehow? I'm a cpp newbie, and https://en.cppreference.com/w/cpp/language/lookup doesn't really explain how namespaces interact with unqualified lookups, but I'd assume it searches the local namespace first (`fmt`, in this case). I thought the whole point of namespaces was avoid these sorts of issues.

Any suggestions / pointers on a fix?

**Comment by a16bitsysop** [comment]
I tried adding this patch:

-```
-- a/src/common/LogEntry.h
+++ b/src/common/LogEntry.h
@@ -214,7 +214,7 @@
 template <> struct formatter<LogEntry> : formatter<std::string_view> {
   template <typename FormatContext>
   auto format(const LogEntry& e, FormatContext& ctx) const {
-    return format_to(ctx.out(), "{} {} ({}) {} : {} {} {}",
+    return fmt::format_to(ctx.out(), "{} {} ({}) {} : {} {} {}",
 			  e.stamp, e.name, e.rank, e.seq, e.channel, e.prio, e.msg);
   }
 };
```

Now I get one error:

```
ninja: job failed: /usr/bin/clang++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DBOOST_MPL_CFG_NO_PREPROCESSED_HEADERS -DBOOST_MPL_LIMIT_LIST_SIZE=30 -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/build/src/include -I/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src -I/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/dmclock/src -I/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/dmclock/support/src -isystem /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/build/include -isystem /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/xxHash -isystem /usr/include/fuse -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1  -D_LARGEFILE64_SOURCE -O2 -DNDEBUG -DSTATX_INO=0x100U -Wno-deprecated-declarations -Os -DNDEBUG -std=c++20 -fPIC -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -fdiagnostics-color=auto -MD -MT src/osd/CMakeFiles/osd.dir/scrubber/scrub_backend.cc.o -MF src/osd/CMakeFiles/osd.dir/scrubber/scrub_backend.cc.o.d -o src/osd/CMakeFiles/osd.dir/scrubber/scrub_backend.cc.o -c /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/osd/scrubber/scrub_backend.cc
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/osd/scrubber/scrub_backend.cc:4:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/osd/scrubber/./scrub_backend.h:50:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/osd/osd_types_fmt.h:8:
In file included from /builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/common/hobject_fmt.h:9:
/usr/include/fmt/ranges.h:314:22: error: no matching member function for call to 'format'
    ctx.advance_to(f.format(v, ctx));
                   ~~^~~~~~
/usr/include/fmt/ranges.h:246:35: note: in instantiation of function template specialization 'fmt::detail::format_tuple_element<fmt::basic_format_context<fmt::appender, char>>::operator()<chunk_info_t>' requested here
  const int unused[] = {0, ((void)f(get<Is>(t1), get<Is>(t2)), 0)...};
                                  ^
/usr/include/fmt/ranges.h:252:3: note: in instantiation of function template specialization 'fmt::detail::for_each2<const std::tuple<fmt::formatter<unsigned long>, fmt::formatter<chunk_info_t>> &, const std::pair<const unsigned long, chunk_info_t> &, fmt::detail::format_tuple_element<fmt::basic_format_context<fmt::appender, char>>, 0UL, 1UL>' requested here
  for_each2(tuple_index_sequence<remove_cvref_t<Tuple1>>(),
  ^
/usr/include/fmt/ranges.h:375:13: note: in instantiation of function template specialization 'fmt::detail::for_each2<const std::tuple<fmt::formatter<unsigned long>, fmt::formatter<chunk_info_t>> &, const std::pair<const unsigned long, chunk_info_t> &, fmt::detail::format_tuple_element<fmt::basic_format_context<fmt::appender, char>>>' requested here
    detail::for_each2(
            ^
/usr/include/fmt/ranges.h:489:25: note: in instantiation of function template specialization 'fmt::formatter<std::pair<const unsigned long, chunk_info_t>>::format<fmt::basic_format_context<fmt::appender, char>>' requested here
      out = underlying_.format(mapper.map(*it), ctx);
                        ^
/usr/include/fmt/ranges.h:548:24: note: in instantiation of function template specialization 'fmt::range_formatter<std::pair<const unsigned long, chunk_info_t>, char>::format<const std::map<unsigned long, chunk_info_t> &, fmt::basic_format_context<fmt::appender, char>>' requested here
    return underlying_.format(range, ctx);
                       ^
/usr/include/fmt/core.h:1308:22: note: in instantiation of function template specialization 'fmt::detail::range_default_formatter<fmt::range_format::map, std::map<unsigned long, chunk_info_t>, char>::format<fmt::basic_format_context<fmt::appender, char>>' requested here
    ctx.advance_to(f.format(*static_cast<qualified_type*>(arg), ctx));
                     ^
/usr/include/fmt/core.h:1291:21: note: in instantiation of function template specialization 'fmt::detail::value<fmt::basic_format_context<fmt::appender, char>>::format_custom_arg<std::map<unsigned long, chunk_info_t>, fmt::formatter<std::map<unsigned long, chunk_info_t>>>' requested here
    custom.format = format_custom_arg<
                    ^
/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/osd/osd_types_fmt.h:79:12: note: in instantiation of function template specialization 'fmt::format_to<fmt::appender, const std::map<unsigned long, chunk_info_t> &, 0>' requested here
      fmt::format_to(ctx.out(), " {}", om.chunk_map);
           ^
/builds/a16bitsysop/aports/community/ceph18/src/ceph-18.2.0/src/osd/osd_types_fmt.h:60:8: note: candidate function template not viable: 'this' argument has type 'const formatter<chunk_info_t, char_type>' (aka 'const formatter<chunk_info_t, char>'), but method is not marked const
  auto format(const chunk_info_t& ci, FormatContext& ctx)
       ^
1 error generated.
```

I am a cpp newbie as well, so I can’t really advise the correct way to fix it.

**Comment by a16bitsysop** [comment]
sorry I think it was me missing previous fmt fixes when I was testing as I haven’t found any fmt fixes in previous searches

**Comment by a16bitsysop** [comment]
I have now tried from the same git commit as your pull request in the alpine CI with your 4 commits and it gets the same error building with fmt 10.1.1:

```
In file included from /builds/a16bitsysop/aports/testing/ceph-git/src/ceph-73dd77f8e74a94ccbaa3478cb2565c811d23b2fe/src/osd/osd_types.cc:43:
In file included from /builds/a16bitsysop/aports/testing/ceph-git/src/ceph-73dd77f8e74a94ccbaa3478cb2565c811d23b2fe/src/osd/osd_types_fmt.h:8:
In file included from /builds/a16bitsysop/aports/testing/ceph-git/src/ceph-73dd77f8e74a94ccbaa3478cb2565c811d23b2fe/src/common/hobject_fmt.h:9:
/usr/include/fmt/ranges.h:314:22: error: no matching member function for call to 'format'
    ctx.advance_to(f.format(v, ctx));
                   ~~^~~~~~
/usr/include/fmt/ranges.h:246:35: note: in instantiation of function template specialization 'fmt::detail::format_tuple_element<fmt::basic_format_context<fmt::appender, char>>::operator()<hobject_t>' requested here
  const int unused[] = {0, ((void)f(get<Is>(t1), get<Is>(t2)), 0)...};
                                  ^
/usr/include/fmt/ranges.h:252:3: note: in instantiation of function template specialization 'fmt::detail::for_each2<const std::tuple<fmt::formatter<hobject_t>, fmt::formatter<interval_set<unsigned long>>> &, const std::pair<const hobject_t, interval_set<unsigned long>> &, fmt::detail::format_tuple_element<fmt::basic_format_context<fmt::appender, char>>, 0UL, 1UL>' requested here
  for_each2(tuple_index_sequence<remove_cvref_t<Tuple1>>(),
  ^
/usr/include/fmt/ranges.h:375:13: note: in instantiation of function template specialization 'fmt::detail::for_each2<const std::tuple<fmt::formatter<hobject_t>, fmt::formatter<interval_set<unsigned long>>> &, const std::pair<const hobject_t, interval_set<unsigned long>> &, fmt::detail::format_tuple_element<fmt::basic_format_context<fmt::appender, char>>>' requested here
    detail::for_each2(
            ^
/usr/include/fmt/ranges.h:489:25: note: in instantiation of function template specialization 'fmt::formatter<std::pair<const hobject_t, interval_set<unsigned long>>>::format<fmt::basic_format_context<fmt::appender, char>>' requested here
      out = underlying_.format(mapper.map(*it), ctx);
                        ^
/usr/include/fmt/ranges.h:548:24: note: in instantiation of function template specialization 'fmt::range_formatter<std::pair<const hobject_t, interval_set<unsigned long>>, char>::format<const std::map<hobject_t, interval_set<unsigned long>> &, fmt::basic_format_context<fmt::appender, char>>' requested here
    return underlying_.format(range, ctx);
                       ^
/usr/include/fmt/core.h:1308:22: note: in instantiation of function template specialization 'fmt::detail::range_default_formatter<fmt::range_format::map, std::map<hobject_t, interval_set<unsigned long>>, char>::format<fmt::basic_format_context<fmt::appender, char>>' requested here
    ctx.advance_to(f.format(*static_cast<qualified_type*>(arg), ctx));
                     ^
/usr/include/fmt/core.h:1291:21: note: in instantiation of function template specialization 'fmt::detail::value<fmt::basic_format_context<fmt::appender, char>>::format_custom_arg<std::map<hobject_t, interval_set<unsigned long>>, fmt::formatter<std::map<hobject_t, interval_set<unsigned long>>>>' requested here
    custom.format = format_custom_arg<
                    ^
/builds/a16bitsysop/aports/testing/ceph-git/src/ceph-73dd77f8e74a94ccbaa3478cb2565c811d23b2fe/src/osd/osd_types.cc:6699:15: note: in instantiation of function template specialization 'fmt::format<const hobject_t &, const eversion_t &, const unsigned long &, const interval_set<unsigned long> &, const std::map<hobject_t, interval_set<unsigned long>> &, const SnapSet &, const bool &>' requested here
  return fmt::format(
              ^
/builds/a16bitsysop/aports/testing/ceph-git/src/ceph-73dd77f8e74a94ccbaa3478cb2565c811d23b2fe/src/common/hobject_fmt.h:33:42: note: candidate function template not viable: 'this' argument has type 'const formatter<hobject_t, char_type>' (aka 'const formatter<hobject_t, char>'), but method is not marked const
  template <typename FormatContext> auto format(const hobject_t& ho, FormatContext& ctx)
                                         ^
1 error generated.
```

Full pipeline: https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/1136215

**Comment by bazaah** [comment]
I'm not really sure what there is for me to do. I don't even touch `osd_types_fmt.h` or `hobject_fmt.h` in this PR... does ceph@main build successfully on alpine without this patch set?

**Comment by a16bitsysop** [comment]
I tried it with adding a patch at a time:
With the first patch https://github.com/ceph/ceph/pull/53698/commits/cc4b87f5c2ee77eb66d25c7b74b29f831cc787f0
It compiles on alpine with fmt 9.1.0 https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/1138710

With the first patch and the second patch https://github.com/ceph/ceph/pull/53698/commits/071c88389f9f4901488095f6efaa7d71982bb5b6
It fails https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/1138719

With the first and third patch https://github.com/ceph/ceph/pull/53698/commits/b20a81686df568c14375083a6a7b8b8dd9d95e55
It passes: https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/1138791

With the second and fourth patch [src: ifdef ostream_formatter impls to >= fmtlib 9](https://github.com/ceph/ceph/pull/53698/commits/3906b34990877906acf756da443efcd46c3cbb2b) depends on [src/common: wrap LogEntry.h fmt impls in namespace](https://github.com/ceph/ceph/pull/53698/commits/071c88389f9f4901488095f6efaa7d71982bb5b6) already being applied so its hard to dissect further

File: src/common/LogEntry.h [hunk 4]
```diff
@@ -15,7 +15,11 @@
 #ifndef CEPH_LOGENTRY_H
 #define CEPH_LOGENTRY_H
 
+#include <fmt/core.h>
```
Comment (tchaikov) [inline src/common/LogEntry.h:18]: why? `fmt/core.h` is already included by, i think, all of the `fmt/*.h` headers? or do you want to explicitly include the very header which has the `FMT_VERSION` macro definition?

File: src/include/byteorder.h [hunk 5]
```diff
@@ -53,3 +53,8 @@ inline ceph_les16 init_les16(__s16 x) {
   v = x;
   return v;
 }
+
+template <typename T>
+auto format_as(ceph_le<T> c) {
+  return (T)c;
```
Comment (tchaikov) [inline src/include/byteorder.h:59]:
```suggestion
  return static_cast<T>(c);
```

probably it's but personal taste, but i'd suggest use a named cast just to be more explicit. see http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named .

File: src/osd/SnapMapper.cc [hunk 6]
```diff
@@ -211,7 +211,7 @@ string SnapMapper::get_prefix(int64_t pool, snapid_t snap)
   return fmt::sprintf("%s%lld_%.16X_",
 		      MAPPING_PREFIX,
 		      pool,
-		      snap);
+		      (uint64_t)snap);
```
Comment (tchaikov) [inline src/osd/SnapMapper.cc:214]: ditto.

File: src/osd/SnapMapper.cc [hunk 7]
```diff
@@ -650,7 +650,7 @@ string SnapMapper::make_purged_snap_key(int64_t pool, snapid_t last)
   return fmt::sprintf("%s_%lld_%016llx",
 		      PURGED_SNAP_PREFIX,
 		      pool,
-		      last);
+		      (uint64_t)last);
```
Comment (tchaikov) [inline src/osd/SnapMapper.cc:653]: ditto.

File: src/osd/SnapMapper.cc [hunk 8]
```diff
@@ -866,7 +866,7 @@ string SnapMapper::get_legacy_prefix(snapid_t snap)
 {
   return fmt::sprintf("%s%.16X_",
 		      LEGACY_MAPPING_PREFIX,
-		      snap);
+		      (uint64_t)snap);
```
Comment (tchaikov) [inline src/osd/SnapMapper.cc:869]: ditto.

File: src/common/LogEntry.h (hunk 4 above)
Comment (bazaah) [reply src/common/LogEntry.h:18]: This was a "copy the documentation thing", happy to remove it

Comment (bazaah) [reply src/common/LogEntry.h]: c3e05b5732af6a0722591f59b991178de3042e45

File: src/include/byteorder.h (hunk 5 above)
Comment (bazaah) [reply src/include/byteorder.h]: c3e05b5732af6a0722591f59b991178de3042e45

**Comment by bazaah** [comment]
Rebased, and adopted feedback into c3e05b5732af6a0722591f59b991178de3042e45. Waiting for test suite

**Pull request closed without merging**